Add basic write support

pka opened the pull request:
This is a minimal writer implementation. Before implementing the missing parts, I would like to discuss the basic design.

Main questions:
1. Should we prepare for other backends (e.g. S3)?
   - (Trait could be done later)
2. Do we want an async API?
   - Probably depends on 1) 

Missing functionality:
* [x] Leaf directories
* [x] API for setting header data (center_zoom, etc.)
* [x] Configurable compression
* [x] Compression for tiles
* [x] Proper errors
* [x] Some TODOs and unwraps

pka committed:
Add basic write support

nyurik commented:
wrt 1 - you mean writing directly to s3?  Does s3 allow random access of the already open file (we would need to go back to the directory to update file positions i would think)

pka commented:
AFAIK, S3 doesn't support seek operations or partial updates, but with a [multipart upload](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpu-upload-object.html), you can defer uploading header and root directory. So a direct upload to S3 would be possble. But this has no high priority for me.

nyurik commented:
TIL, thx.  And lets follow [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it) - if someone wants it, they will add it :)

pka committed:
Compare tile bytewise instead of hashing

pka committed:
Ignore empty tiles

pka committed:
Remove __writer feature

pka committed:
Use shared compression function

pka committed:
Build and write leaf directories

pka committed:
Add writer builder

pka committed:
Make writer generig for Write+Seek

pka committed:
Track n_tile_contents

pka committed:
Add usage example

pka commented:
This is now ready for a review.

I realized an important property, which I want to add as a followup: deduplicating non-subsequent tiles. This has performance drawbacks when reading, but probably reduces the file size substantially.

pka committed:
Use builder API in test

nyurik commented on src/writer.rs:
```diff
@@ -0,0 +1,450 @@
+    }
+
+    /// Finish writing the `PMTiles` file.
+    pub fn finish(mut self) -> PmtResult<()> {
```
If finalizing a file is required, should we add the `Drop` support, possibly panicing on error? Otherwise we could be left in a bad non-finalized state... unless of course user ends writing early due to some other error - in which case we are also left in a bad state, but we shouldn't finalize it...   Or perhaps we should delete the output?

We could also introduce `poison` concept - if any `write` method on this struct fail, we set a flag before returning an error, and in that case the file will not be finalized automatically. Otherwise, finalize the file on Drop unless it was already finalized by explicit call.

How does std::fs  handle errors during Drop?

pka replied:
I'm not sure about that. I see the point, but don't think that we should put too much effort into fixing an obvious user bug.

nyurik replied:
ok, lets keep as is.  I tend to use the word `finalize`, but either works

nyurik commented on src/writer.rs:
```diff
@@ -0,0 +1,450 @@
+}
+
+fn compare_blobs(a: &[u8], b: &[u8]) -> bool {
+    a.len() == b.len() && a.iter().zip(b.iter()).all(|(a, b)| a == b)
```
why not `a == b` ?

pka replied:
You are kidding me???

nyurik commented on src/writer.rs:
```diff
@@ -0,0 +1,450 @@
+    a.len() == b.len() && a.iter().zip(b.iter()).all(|(a, b)| a == b)
+}
+
+fn into_u32(v: usize) -> PmtResult<u32> {
```
it seems this is only used twice - should it be a func?  I'm ok either way

pka replied:
Maybe it also useful for the reader, where currently `allow(clippy::cast_possible_truncation)` is set.

nyurik reviewed (commented):
looks pretty good!  I left a few thoughts

pka committed:
Minor cleanups

pka committed:
Check tile_id order

pka commented on src/directory.rs:
```diff
@@ -17,6 +19,19 @@ impl Debug for Directory {
+            entries: Vec::with_capacity(capacity),
+        }
+    }
+    pub(crate) fn from_entries(entries: &[DirEntry]) -> Self {
```
There is no case for that yet. It's mainly used for temporary directories for optimizing leaves.

nyurik replied:
let me rephrase - per Rust recommended usage, it is always better to "ask for ownership if you need to own it" - this way the caller will have the choice of cloning or passing the value's ownership if no longer needed. Here this is an internal API, so it is up to you.

pka commented on src/lib.rs:
```diff
@@ -1,3 +1,19 @@
+//! Read and write `PMTiles` according to the [PMTiles v3 spec](https://github.com/protomaps/PMTiles/blob/master/spec/v3/spec.md).
```
I'm fine with both. Do the doc checks also run, when README.md is included?

pka commented on src/writer.rs:
```diff
@@ -0,0 +1,450 @@
+            min_zoom: 0,
+            max_zoom: 22,
+            min_longitude: -180.0,
+            min_latitude: -85.0,
```
Thanks for the hint. I took the numbers from the reference example, but your figure is more precise and the same as used by MapLibre JS.

pka commented on src/writer.rs:
```diff
@@ -0,0 +1,450 @@
+impl<W: Write + Seek> PmTilesStreamWriter<W> {
+    /// Add tile to writer
+    /// Tiles are deduplicated and written to output.
+    /// `tile_id` should be increasing.
```
Adressed in 4586c4c

pka commented:
Regarding the failed semver check:

```
--- failure enum_variant_added: enum variant added on exhaustive enum ---

Description:
A publicly-visible enum without #[non_exhaustive] has a new variant.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#enum-variant-new
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/enum_variant_added.ron

Failed in:

     Summary semver requires new major version: 1 major and 0 minor checks failed
  variant PmtError:IndexEntryOverflow in /home/runner/work/pmtiles-rs/pmtiles-rs/src/error.rs:28
```

Shall I add `#[non_exhaustive]` or increase the version?

pka committed:
Move usage example to README an include it in module docs

pka committed:
Rename finish to finalize

pka committed:
Directory::from_entries now takes ownership

pka committed:
Export tile_id function

nyurik commented:
Might as well bump the version

pka committed:
Make PmtError non_exhaustive and bump version

nyurik commented:
Why do you want to make it non-exhaustive? I am ok to break compat - not a biggie tbh

nyurik commented:
but come to think of it, errors are a pretty good candidate for non-exhaustive, so yeah, lets keep it going forward... but enabling it would still be a breaking change, right?

nyurik commented:
one more thing - i think `write` support is a less used feature, and since it requires extra dependencies, lets make this into an optional feature

pka commented:
> one more thing - i think `write` support is a less used feature, and since it requires extra dependencies, lets make this into an optional feature

`flate2` is already an indirect dependency and `countio` is quite small. So I thougt it's not worth the hassle.

lseelenbinder commented:
Ohhh! This is great. Just seeing the PR now for some reason. I'll give it a review shortly. Thank you @pka!

pka committed:
Change gzip compression level to 6

About 3 times faster and very similar compression rate for MVTs

nyurik commented on README.md:
```diff
@@ -17,16 +17,31 @@ originally created by Brandon Liu for Protomaps.
 
+## Usage examples
+
+### Writing a PMTiles file
```
We must add a read example here before merging, or else it will get really confusing why we have a write without read

nyurik replied:
Come to think of it - agree, this is not a blocker for merging

nyurik commented on README.md:
```diff
@@ -17,16 +17,31 @@ originally created by Brandon Liu for Protomaps.
+
+let file = File::create("tiles.pmtiles").unwrap();
+let mut writer = PmTilesWriter::new(TileType::Mvt).create(file).unwrap();
+writer.add_tile(tile_id(0, 0, 0), &[/*...*/]).unwrap();
```
do we allow zero-length tiles?

pka replied:
Zero-length tiles added with `add_tile` are ignored. The spec doesn't allow storing zero-length tiles and I think throwing an error is not absolutely necessary.

lseelenbinder replied:
I agree.

nyurik commented on src/lib.rs:
```diff
@@ -15,8 +17,8 @@ pub mod cache;
 mod header;
-#[cfg(feature = "__async")]
 mod tile;
+mod writer;
```
could you clarify why cfg(__async) was removed here?  We really ought to make this PR behind a feature flag

pka replied:
I was glad to reduce the `cfg(__)` complexity a bit with 72f57629 - it reminded me of the bad old C/++ `#ifndef` days. IMO, a pmtile library should come with its core functionality (like calculating a tile id) and dependencies (like flate2) by default. I understand the feature flags for the different backend implementations with heavy or specialized dependencies, but not for everything. That said, if the maintainers prefer a feature flag for writing pmtile files, I can give it a try.

nyurik replied:
I agree that we shouldn't go overboard on features.  My only concern is that if we have additional dependencies due to write support, it is better to make write - optional, simply because write is a relatively rare usage compare to read. My understanding is that you need compression, which might be a significant extra compilation burden - and that IMO should not be levied on users unless needed.

lseelenbinder reviewed (changes requested):
First of all, thanks for setting this all up! I'm glad we're getting this functionality added.

I'd like to understand how much impact having the leaf directories at the end of the archive would have on performance. Hopefully it's not too much, because this elegantly solves the need for a temporary file.

Another option is to reserve N bytes at the beginning of every archive for the header + directories, but that's suboptimal as well.

pka commented:
> I'd like to understand how much impact having the leaf directories at the end of the archive would have on performance. Hopefully it's not too much, because this elegantly solves the need for a temporary file.
> 
> Another option is to reserve N bytes at the beginning of every archive for the header + directories, but that's suboptimal as well.

I would not expect a performance impact, even via http. An optimization for initial loading would be adding the entries for z0-z2 to the root directory, which is a TODO copied from the pmtiles-go implementation. Shouldn't be difficult to add to the current implementation.

pka commented on src/writer.rs:
```diff
@@ -0,0 +1,492 @@
+            n_tile_entries: None,
+            n_tile_contents: None,
+            clustered: true,
+            internal_compression: Compression::Gzip,
```
That's the default value which can be overridden with the `internal_compression` builder function. But currently, only compression methods supported by the reader are included.

nyurik replied:
on naming - `internal` is somewhat confusing - is this the header and/or metadata compression?

pka replied:
Thats the spec naming: https://github.com/protomaps/PMTiles/blob/main/spec/v3/spec.md#internal-compression-ic

pka commented on src/writer.rs:
```diff
@@ -0,0 +1,492 @@
+
+    /// Set the compression for metadata and directories.
+    #[must_use]
+    pub fn internal_compression(mut self, compression: Compression) -> Self {
```
`PmTilesWriter` *is* a builder struct, generating a `PmTilesStreamWriter` struct with `create`. `Builder` is not in the name, because the resulting code looks better without it, IMO.

lseelenbinder replied:
If we switched to `pmtiles::writer::StreamWriter` then we could have `pmtiles::writer::Builder` without confusion.

lseelenbinder replied:
I _think_ we can accomplish the same thing with name spacing: `pmtiles::writer::StreamBuilder` and `pmtiles::writer::StreamWriter`. It's a pretty common pattern in Rust, and I think keeps things concise and manageable.

Any thoughts on this @nyurik?

nyurik replied:
It took me a few seconds to understand "name spacing" meant "namespacing" :D  Naming is hard...

I am not a big fan of mega-generic names like `Builder` by itself, nor adding namespace as part of the required pattern... I do like `PmTilesWriter::new(...).create(...)` as it is cleaner to the user.  We could name the resulting struct as `PmTilesWriterStream` or `PmTilesStream` or whatever.

lseelenbinder replied:
Okay, this is a pretty fundamental decision—so I'd say this blocks the MR until we decide.

I did some research, and here's my suggestion.

The "thing" we want to build is called `PmTilesWriter` (I think "stream" is an implementation—99% of use-cases prefer streaming anyways.).

Then we can have a `Builder` in the same module, but hide the complexity behind a `PmTilesWriter::builder()` call.

This results in the following pseudocode:

```rust
let writer = PmTilesWriter::builder(tile_type).create("file.pmtiles");
```

The user never needs to know about the underlying builder structure, and it keeps the naming manageable. I'm happy to do the refactor, but just want to make sure there's no strong objections.

nyurik replied:
@lseelenbinder I just realized you were waiting for me here - this looks fine, want to make that change?  Lets merge and push this!

pka commented on src/writer.rs:
```diff
@@ -0,0 +1,492 @@
+    /// Add tile to writer.
+    ///
+    /// Tiles are deduplicated and written to output.
+    /// `tile_id` should be increasing for best read performance.
```
In contrary to the https://github.com/arma-place/pmtiles-rs writer, tiles are written directly to disk an not held in memory and sorted before writing. So `add_tile` expects the tiles already in correct order, which means `tile_id` has to be known anyway.

lseelenbinder replied:
Ah, got it. I was mixing up `tile_id` with something else—forgetting it's just the (z,x,y) on a hilbert curve.

pka commented on src/writer.rs:
```diff
@@ -0,0 +1,492 @@
+        &self,
+        writer: &mut Counter<W>,
+        compression: Compression,
+    ) -> PmtResult<usize> {
```
In consequence, the reader(s) should then also have its own error enum(s).

lseelenbinder replied:
That isn't a bad thing, in my opinion. The Error type is quite packed right now.

I think `OpenError`, `ReadError`, and `WriteError` with associated `Result` types makes sense at this point, and leads to better code and a clearer API. @nyurik did a lot of the Error-related work, so his input is appreciated.

lseelenbinder commented on src/writer.rs:
```diff
@@ -0,0 +1,492 @@
+        let mut out = Counter::new(BufWriter::new(writer));
+
+        // We use the following layout:
+        // +--------+----------------+----------+-----------+------------------+
+        // |        |                |          |           |                  |
+        // | Header | Root Directory | Metadata | Tile Data | Leaf Directories |
+        // |        |                |          |           |                  |
+        // +--------+----------------+----------+-----------+------------------+
+        // This allows writing without temporary files. But it requires Seek support.
```
I see. I think you're right. The root directory is usually read as part of the header, and then the leaf directories are clustered, just as they would be normally.

This is a pretty clever solution, to be honest. I'm curious what @bdon thinks about this for go-pmtiles?

bdon replied:
leaf directories at the end is in-spec and also how the java implementation in planetiler works: 

https://github.com/onthegomap/planetiler/blob/main/planetiler-core/src/main/java/com/onthegomap/planetiler/pmtiles/WriteablePmtiles.java#L220

lseelenbinder commented:
> wrt 1 - you mean writing directly to s3? Does s3 allow random access of the already open file (we would need to go back to the directory to update file positions i would think)

I think this would be theoretically possible if we use multi-part upload, and refrain from uploading the first part until the end, but I'd have to check the docs to confirm this is possible.

lseelenbinder commented:
When I have the chance, I was planning to grab a copy of this PR and do some additional cleanup that makes sense now. I'll make it dependent against your code, @pka, so we don't have merge issues.

I'm struggling with how much to ask for in this PR vs what we should clean up after. Let's *not* immediately release after merging to give us some time to clean up the APIs in a more general way.

nyurik commented on src/writer.rs:
```diff
@@ -0,0 +1,492 @@
+    #[must_use]
+    pub fn new(tile_type: TileType) -> Self {
+        let tile_compression = match tile_type {
+            TileType::Mvt => Compression::Gzip,
```
I do too, but let's limit the scope of this PR :)

nyurik commented on src/writer.rs:
```diff
@@ -0,0 +1,492 @@
+    header: Header,
+    entries: Vec<DirEntry>,
+    n_addressed_tiles: u64,
+    // TODO: Replace with digest HashMap for deduplicating non-subsequent tiles
```
I agree, but lets try to get this merged with MVP functionality first, or else we risk dragging this PR forever :)

lseelenbinder replied:
Agreed.

pka committed:
Add links to reference implementations

pka commented on src/writer.rs:
```diff
@@ -0,0 +1,492 @@
+    }
+
+    fn optimize_directories(&self, target_root_len: usize) -> PmtResult<(Directory, usize)> {
+        // Same logic as go-pmtiles optimizeDirectories
```
Links added.

lseelenbinder commented on src/writer.rs:
```diff
@@ -0,0 +1,492 @@
+    ///
+    /// Tiles are deduplicated and written to output.
+    /// `tile_id` should be increasing for best read performance.
+    pub fn add_tile(&mut self, tile_id: u64, data: &[u8]) -> PmtResult<()> {
```
Let's make this a TODO for after merging.

lseelenbinder commented:
Okay! We're close. Just have to resolve the naming detail, and I'll be happy to do a cleanup pass after merging.

nyurik committed:
chore: more justfile updates

* in CI, ensure checkout happens before caching; otherwise it won't work
* use json-based retrieval of metadata
* fix `just test-all` to use the right fmt
* fix `fmt` to properly test if fmt nightly is there

nyurik committed:
consolidate all compat features, fix docs

nyurik committed:
Merge branch 'main' into writer

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nyurik committed:
just fmt

nyurik committed:
Merge branch 'just' into writer

nyurik committed:
make write optional

nyurik committed:
don't add fmt and clippy automagically

it is not worth it because these just commands are only for local development, and most users install rust with the default profile which includes both. On the other hand, constant adding them in every call slows things down and may hide other bigger issues with the local installation.

nyurik committed:
Merge branch 'just' into writer

nyurik commented:
ok, lots of changes here - write is now feature-gated, lots of other minor cleanups and changes.  This PR relies on #56, but once merged, should be easier to review.

nyurik committed:
rename internal func, bump

nyurik committed:
Merge branch 'just' into writer

nyurik committed:
nits

nyurik committed:
Add readme docs

nyurik committed:
Merge branch 'just' into writer

nyurik committed:
lints

nyurik committed:
Merge branch 'just' into writer

nyurik committed:
rm redundant test, thx ai

nyurik committed:
Merge branch 'main' into writer

nyurik committed:
move header creation to new()

nyurik commented on src/writer.rs:
```diff
@@ -0,0 +1,492 @@
+            _ => Compression::None,
+        };
+        #[allow(clippy::excessive_precision, clippy::unreadable_literal)]
+        let header = Header {
```
i made it into a `new` with a few params

nyurik reviewed (commented):
@lseelenbinder i made a few more refactorings. Any blockers to merge this?

nyurik committed:
grammmar

lseelenbinder reviewed (approved):
LGTM!

I think there's quite a few low-hanging improvements for future PRs, but this looks solid as a first release.
